### PR TITLE
[miri] Throw UB if target size and data size don't match

### PR DESCRIPTION
**Pull request by samrat**
Issue: https://github.com/rust-lang/miri/issues/1355

If an extern C function is defined as

```
extern "C" {
    fn malloc(size: u32) -> *mut std::ffi::c_void;
}
```

on a 64-bit machine(ie. pointer sizes don't match), return undefined behaviour from Miri when [converting the argument into machine_usize](https://github.com/rust-lang/miri/blob/master/src/shims/foreign_items.rs#L200)

### CODE REVIEW

File: src/librustc_middle/mir/interpret/value.rs [hunk 1]
```diff
@@ -393,7 +393,12 @@ impl<'tcx, Tag> Scalar<Tag> {
         assert_ne!(target_size.bytes(), 0, "you should never look at the bits of a ZST");
         match self {
             Scalar::Raw { data, size } => {
-                assert_eq!(target_size.bytes(), u64::from(size));
+                if target_size.bytes() != u64::from(size) {
+                    throw_ub!(ArgumentSizeMismatch {
+                        target_size: target_size.bytes(),
+                        data_size: u64::from(size)
```
Comment (RalfJung) [inline src/librustc_middle/mir/interpret/value.rs:399]:
```suggestion
                        data_size: u64::from(size),
```

File: src/librustc_middle/mir/interpret/error.rs [hunk 2]
```diff
@@ -422,6 +427,11 @@ impl fmt::Debug for UndefinedBehaviorInfo {
                 "using uninitialized data, but this operation requires initialized memory"
             ),
             DeadLocal => write!(f, "accessing a dead local variable"),
+            ArgumentSizeMismatch { target_size, data_size } => write!(
```
Comment (RalfJung) [inline src/librustc_middle/mir/interpret/error.rs:430]: This function is not just used for arguments, it is used for scalars occurring anywhere in the program. Thus, I think `ScalarSizeMismatch` is a better name.

**Comment by RalfJung** [comment]
@samrat thanks a lot, this looks great! I just have some minor nits.

File: src/librustc_middle/mir/interpret/error.rs (hunk 2 above)
Comment (samrat) [reply src/librustc_middle/mir/interpret/error.rs]: :+1: Changed this

File: src/librustc_middle/mir/interpret/error.rs [hunk 3]
```diff
@@ -361,6 +361,11 @@ pub enum UndefinedBehaviorInfo {
     InvalidUndefBytes(Option<Pointer>),
     /// Working with a local that is not currently live.
     DeadLocal,
+    /// Data size is not equal to target size
```
Comment (RalfJung) [inline src/librustc_middle/mir/interpret/error.rs:364]:
```suggestion
    /// Data size is not equal to target size.
```

**Comment by RalfJung** [comment]
I adjusted style a bit more as I had missed that before. Thanks!

@bors r+ rollup

**Pull request merged**